semgrep-core/Parsing/other/yaml_to_generic.ml

```diff
@@ -0,0 +1,121 @@
+
+let make_scalar _anchor _tag pos value file : A.expr =
+  A.L (A.String (value, mk_tok pos value file))
+
+(* Sequences are arrays in the generic AST *)
+let make_sequence _anchor _tag start_pos (es, end_pos) file =
+  A.Container(A.Array, mk_bracket (start_pos, end_pos) es file)
+
+(* Mappings are dictionaries in the generic AST *)
+let make_mappings _anchor _tag start_pos (es, end_pos) file =
+  A.Container(A.Dict, mk_bracket (start_pos, end_pos) es file)
+
+let parse file parser : A.any =
+  (* Parse states *)
+  let rec read_stream () : A.any =
+    match get_res (S.do_parse parser) with
+    | E.Stream_start _, _pos -> any_of_list (read_documents [])
+    | v, pos -> raise (ParseError (mk_err "Expected start of string, got" v pos))
+  and read_documents acc : A.expr list =
+    match get_res (S.do_parse parser) with
+    | E.Document_start _, _pos -> let docs = read_document () in docs :: (read_documents acc)
+    | E.Stream_end, _ -> acc
+    | v, pos -> raise (ParseError (mk_err "Expected start of document or end of stream, got" v pos))
+  and read_document () : A.expr =
+    let node_ast, _ = read_node () in
+    match get_res (S.do_parse parser) with
+    | E.Document_end _, _pos -> node_ast
+    | v, pos -> raise (ParseError (mk_err "Expected end of document, got" v pos))
+  and read_node ?node_val:(res = None) () : A.expr * E.pos =
+    let res = match res with None -> get_res (S.do_parse parser) | Some r -> r in
+    match res with
+    | E.Alias { anchor }, pos -> A.N (make_alias anchor pos file), pos
+    | E.Scalar { anchor; tag; value; _ }, pos -> make_scalar anchor tag pos value file, pos
+    | E.Sequence_start { anchor; tag; _ }, pos -> make_sequence anchor tag pos (read_sequence []) file, pos
+    | E.Mapping_start { anchor; tag; _ }, pos -> make_mappings anchor tag pos (read_mappings []) file, pos
+    | v, pos -> raise (ParseError (mk_err "Expected a valid YAML element or end of sequence, got" v pos))
+  and read_sequence acc : A.expr list * E.pos =
+    match get_res (S.do_parse parser) with
+    | E.Sequence_end, pos -> acc, pos
+    | v, pos -> let seq = (fst (read_node ~node_val:(Some(v, pos)) ())) in seq :: (fst (read_sequence acc)), pos
+  and read_mappings acc : A.expr list * E.pos =
+    match get_res (S.do_parse parser) with
+    | E.Mapping_end, pos -> acc, pos
+    | v, pos -> let map = read_mapping (v, pos) in map :: (fst (read_mappings acc)), pos
+  and read_mapping first_node : A.expr =
+    let key, pos1 =
+      match first_node with
+      | E.Scalar { anchor; tag; value; _ }, pos -> make_scalar anchor tag pos value file, pos
+      | v, pos -> raise (ParseError (mk_err "Expected a valid scalar, got" v pos))
+    in
+    let value, pos2 = read_node () in
+    A.Tuple (mk_bracket (pos1, pos2) [key; value] file)
+  in
+  read_stream ()
+
+let parse_yaml str =
+  get_res (S.parser str) |> parse "dummy_file_name"
+
+(* Entry point *)
+
+let program _file = (* parse_yaml file *) raise Unimplemented
```

Comment:
Why not implemented?
We actually have a file here, so we can have the right position information.
The position information for a pattern is not that important; it's really important for the yaml target file though,
because we report ranges on those target file.

BTW you may need to pass down more information than a file later, so it's good practice to define
type env = {
  file: Common.filename;
}
and pass that around instead of just the file.
That way you can add more info later.